.github/workflows/ci-cd.yml

```diff
@@ -0,0 +1,187 @@
+name: CI/CD
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+env:
+  NODE_VERSION: 10
+
+jobs:
+  lint:
+    name: Lint files
+    runs-on: ubuntu-latest
+    timeout-minutes: 7
+    steps:
+      - name: Check out a copy of the repo
+        uses: actions/checkout@v2
+
+      - name: Use Node.js ${{ env.NODE_VERSION }}
+        uses: actions/setup-node@v2-beta
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+
+      - name: Get Yarn cache path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+
+      - name: Cache Yarn cache and node_modules
+        id: cache-dependencies
+        uses: actions/cache@v2
+        with:
+          path: |
+            ${{ steps.yarn-cache-dir-path.outputs.dir }}
+            node_modules
+            test-apps/new-addon/node_modules
+          key: ${{ runner.os }}-${{ env.NODE_VERSION }}-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: ${{ runner.os }}-${{ env.NODE_VERSION }}-
+
+      - name: Install dependencies
+        run: yarn install --frozen-lockfile
+        if: steps.cache-dependencies.outputs.cache-hit != 'true'
+
+      - name: Lint
+        run: yarn lint
+
+
+  test-app:
+    name: Test app
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        script-name:
+          - 'ember'
+          - 'node'
+          - 'test-apps'
+    timeout-minutes: 7
+    steps:
+      - name: Check out a copy of the repo
+        uses: actions/checkout@v2
+
+      - name: Use Node.js ${{ env.NODE_VERSION }}
+        uses: actions/setup-node@v2-beta
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+
+      - name: Get Yarn cache path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+
+      - name: Cache Yarn cache and node_modules
```

Comment:
@ijlee2 I think perhaps we don't want to cache everything every time? I think we previously had some tests for locked and floating deps. Also, it seems most steps do the same steps of checkout, use node, install etc. Can we eliminate some of these installs and use the already installed stuff?

Reply:
I appreciate your feedback! Let me address them in reverse order:

- To my understanding, currently it's not possible to do a "before each" hook (YAML anchor) in GitHub Actions. Having duplicate steps across different jobs is unfortunate but necessary.

    A workaround may be possible with a [composite run steps action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/about-actions#composite-run-steps-actions), but I think it's better to duplicate code (simple approach) than abstract by creating an action (more complex). (There may be an action within the Ember community and I haven't found it yet.)

- I didn't understand why floating dependencies tests were needed for an addon. From the [ember-try documentation](https://github.com/ember-cli/ember-try#a-note-on-npm-scenarios-with-lockfiles), it does seem like we do want to continue to have floating dependencies tests. I can work on updating the workflow to restore this feature.

- For locked dependencies tests (if I understood what locked dependencies tests do versus floating), I think caching makes sense and will help decrease the runtime.

Can you help me understand the locked vs. floating tests better?

Reply:
I do not know personally why locked vs floating matters, but I think it's just to make sure the latest versions of deps are not breaking things, so you can catch it early.

I agree that caching while using locked deps is good 👍 

For skipping some duplicate steps, I think it is possible to use output from a previous step in a subsequent step. I.e. we should be able to choose the node version once, instead of every time, but I do not know the specifics myself.